### PR TITLE
fix: Require `instructlab-sdg>=0.6.2` to resolve dependency issues + require `pydantic<=2.9.2`

### LINKED ISSUE #2765: InstructLab fails to run after `docling-parse` major version upgrade
**Describe the bug**

At 4:35am ET, the `docling-parse` maintainers released a new major version, v3.0.0: https://github.com/DS4SD/docling-parse/releases/tag/v3.0.0

This version contains a number of breaking changes that essentially change some syntax as well as remove the `docling_parse.docling_parse` Python subpackage. Example:

https://github.com/DS4SD/docling-parse/pull/73/files#diff-8c4b81c815639e7a721bfea7a0e5b68828d2695705557b2a282e6647329c218bL43

Because of these changes and because we are not pinning to a specific version of the `docling-parse` package, our small e2e jobs seem to fail with this error:

```python3
  failed to generate data with exception: No module named 'docling_parse.docling_parse'
  +2024-12-09 21:00:06 5.2.37(1)-release:./scripts/functional-tests.sh:341: test_data_generate(): echo 'Error: ilab data generate command took more than 20 minutes and was cancelled'
```

**To Reproduce**
1. Create a pull request
2. Let the pull request e2e jobs run

**Expected behavior**

The e2e tests pass

**Screenshots**

Screenshot 2024-12-09 at 4 27 57 PM

**Device Info (please complete the following information):**
 - Hardware Specs: N/A
 - OS Version: Ubuntu + macOS so far, but likely all OSes due to the nature of the bug
 - Python Version: 3.10 + 3.11
 - InstructLab Version: `main` build

**Additional context**

### LINKED ISSUE #2773: Linter issue: `mypy<1.14` pulls in incompatible dependencies by default
**Describe the bug**

We use [Mypy](https://github.com/python/mypy) to perform lint checks during PR builds (see [this line](https://github.com/instructlab/instructlab/blob/0ccce67d99733d265e7401061c438d438753d6fc/tox.ini#L137) in `tox.ini`). However, `mypy` has started to incorrectly complain about variable type mismatches. (See screenshot in this issue.)

It turns out that one of `mypy`'s dependencies, `pydantic`, was recently updated and `mypy` is no longer fully compatible with `pydantic`'s latest code.  This compatibility issue is problematic because `mypy<1.14` seems to pull in the latest `pydantic` release by default through a dependency chain, which means certain `mypy<1.14` logic no longer works as intended.

Note: I have found two relevant issues filed for both `mypy` and `pydantic`, but there may be more as time marches on:

  - https://github.com/python/mypy/issues/18191
  - https://github.com/pydantic/pydantic/issues/10950

**To Reproduce**
1. Create a pull request that updates Python code
2. Let the `mypy` linter check run

**Expected behavior**

The `mypy` linter check passes

**Screenshots**
Screenshot 2024-12-09 at 5 05 51 PM

**Device Info (please complete the following information):**
 - Hardware Specs: N/A
 - OS Version: N/A
 - Python Version: 3.10 + 3.11
 - InstructLab Version: `main` branch

**Additional context**

N/A

### PR DESCRIPTION
**Pull request by courtneypacheco**
**Issue resolved by this Pull Request:**
Resolves #2765
Resolves #2773

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

In the `docling-parse` v3.0.0 release, the maintainers rearranged a lot of logic and deprecated certain syntaxes, which has resulted in our InstructLab e2e tests failing to build. (See #2765 for more details.) Since `docling-parse` is indirectly pulled in via the `instructlab-sdg` package,  the `docling-parse`/`docling` issues were resolved in `instructlab-sdg` release [v0.6.2](https://github.com/instructlab/sdg/releases/tag/v0.6.2) and I have therefore updated this PR to require `instructlab-sdg>=0.6.2`.

Also, there was another unforeseen issue that this PR fixes: We use `mypy` to execute our linting steps, but one of `mypy`'s dependencies was recently updated with breaking changes:

Screenshot 2024-12-09 at 5 05 51 PM

Doing a quick investigation, I see that `mypy` has some issues filed for this problem:

  - https://github.com/python/mypy/issues/18191
  - https://github.com/pydantic/pydantic/issues/10950

To workaround this dependency issue, I pinned the related dependency, `pydantic`, to `<=v2.9.2` in our `tox.ini` file. This will force our latest `mypy` to use a compatible `pydantic`. However, note that I did also pin `mypy>=1.0,<1.14`. I did this as a safety measure for when the `mypy` maintainers inevitably fix the issue in v1.14 or later.

### CODE REVIEW

**Review by nathan-weinberg — APPROVED** [review]

File: requirements.txt
Comment (cdoern) [reply requirements.txt:6]: could we alternatively just pin ==2.0.0 in order to not pick up any other 2.y releases? or we could do >=2.0.0,<3.0.0

Comment (courtneypacheco) [reply requirements.txt]: I changed it to `docling-parse>=2.0.0,<3.0.0`. 

I thought about pinning directly to v2.0.0, but if the `docling-parse` maintainers push any security patches or major bug fixes for v2, I think we want to automatically consume those -- hence the `docling-parse>=2.0.0,<3.0.0`.

It appears the syntax etc. breaking changes are only effective for >= v3.0.0, so I don't think we have to worry too much about 2.y releases breaking syntax.

Also, these changes in my PR are a temporary measure to unblock PR builds, so we should ultimately update the syntax here to use the new v3 syntax.

File: requirements.txt
Comment (dhellmann) [inline requirements.txt:6]: We have some customizations in the build downstream for docling-parse, but they don't look like anything that would be affected by this and we're currently building version 2.1.2. If 3.0 _is_ significantly breaking, we'll want to coordinate the update so the builder code is ready when upstream switches this to `>=3.0.0`.

File: tox.ini
Comment (dhellmann) [inline tox.ini:137]: Since these are developer tools, only used in CI, it would be safe to use `==` to pick a version and then let dependabot or renovate update the version with a PR. That would keep you up to date, but only allow updates that pass the tests.

**Review by dhellmann — COMMENTED** [review]
This looks good to me. I'll leave it for the core team to approve.

File: tox.ini
Comment (courtneypacheco) [reply tox.ini:137]: Removed and introduced in the `instructlab-sdg` package. See SDG release [v0.6.2](https://github.com/instructlab/sdg/releases/tag/v0.6.2).

File: requirements.txt
Comment (courtneypacheco) [reply requirements.txt]: Removed and introduced in the `instructlab-sdg` package. See SDG release [v0.6.2](https://github.com/instructlab/sdg/releases/tag/v0.6.2).

**Review by danmcp — APPROVED** [review]

**Pull request merged**
